### PR TITLE
Implementation for dpnp.full()

### PR DESCRIPTION
**Pull request by chudur-budur**
Implementation for `dpnp.full()`.

This will enable calling `dpnp.full()` inside `@dpjit` functions.

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Comment by chudur-budur** [comment]
@diptorupd

**Comment by chudur-budur** [comment]
keeping track on #254

File: numba_dpex/core/runtime/_dpexrt_python.c [hunk 1]
```diff
@@ -521,8 +549,9 @@ DPEXRT_MemInfo_alloc(npy_intp size, size_t usm_type, const char *device)
  */
```
Comment (diptorupd) [inline numba_dpex/core/runtime/_dpexrt_python.c:549]: Docstring needs to be updated.

File: numba_dpex/core/runtime/_dpexrt_python.c [hunk 2]
```diff
@@ -553,12 +582,16 @@ static NRT_MemInfo *DPEXRT_MemInfo_fill(NRT_MemInfo *mi,
     case 3:
     {
         uint64_t value_assign = (uint64_t)value;
-        if (is_float) {
-            double const_val = (double)value;
+        if (dest_is_float && !value_is_float) {
+            double const_val = value;
```
Comment (diptorupd) [inline numba_dpex/core/runtime/_dpexrt_python.c:586]: Why not use a union type for double as well?

File: numba_dpex/core/runtime/_dpexrt_python.c [hunk 3]
```diff
@@ -23,6 +23,28 @@
...
+        fflush(stdout);                                                        \
+    }
+#else
+#define DPEXRT_DEBUG(X)                                                        \
+    if (0) {                                                                   \
+        X;                                                                     \
+    }
+#endif
+
+typedef union
```
Comment (diptorupd) [inline numba_dpex/core/runtime/_dpexrt_python.c:42]: Docstring is needed

File: numba_dpex/core/runtime/_dpexrt_python.c [hunk 4]
```diff
@@ -521,8 +549,9 @@ DPEXRT_MemInfo_alloc(npy_intp size, size_t usm_type, const char *device)
  */
 static NRT_MemInfo *DPEXRT_MemInfo_fill(NRT_MemInfo *mi,
                                         size_t itemsize,
-                                        bool is_float,
-                                        uint8_t value,
+                                        bool dest_is_float,
+                                        bool value_is_float,
+                                        int64_t value,
```
Comment (diptorupd) [inline numba_dpex/core/runtime/_dpexrt_python.c:554]: Do we really need `int64_t`? Can you please evaluate with `uint64_t` and using the union for bit-pattern conversion?

File: numba_dpex/core/runtime/context.py [hunk 5]
```diff
@@ -33,10 +33,25 @@ def meminfo_alloc(self, builder, size, usm_type, device):
...
+        self,
+        builder,
+        meminfo,
+        itemsize,
+        dest_is_float,
+        value_is_float,
+        value,
+        device,
+    ):
         """A wrapped caller for meminfo_fill_unchecked() with null check."""
```
Comment (diptorupd) [inline numba_dpex/core/runtime/context.py:46]: The docstring should be improved.

Comment (diptorupd) [reply numba_dpex/core/runtime/context.py:46]: We need a link to `meminfo_fill_unchecked` and properly explain what the null check is and how it works.

File: numba_dpex/core/runtime/context.py [hunk 6]
```diff
@@ -71,7 +86,14 @@ def meminfo_alloc_unchecked(self, builder, size, usm_type, device):
...
+        self,
+        builder,
+        meminfo,
+        itemsize,
+        dest_is_float,
+        value_is_float,
+        value,
+        device,
     ):
         """Fills an allocated `MemInfo` with the value specified.
```
Comment (diptorupd) [inline numba_dpex/core/runtime/context.py:98]: Docstring needs to be updated.

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 7]
```diff
@@ -76,14 +77,25 @@ def fill_arrayobj(context, builder, sig, llargs, value, is_like=False):
         types.intp, get_itemsize(context, arrtype[0])
     )
     device = context.insert_const_string(builder.module, arrtype[0].device)
-    value = context.get_constant(types.int8, value)
+
+    value_is_float = context.get_constant(types.boolean, 0)
+    if isinstance(sig.args[1], types.scalars.Float):
+        value = builder.bitcast(value, llvmir.IntType(64))
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:83]: Can you test if the bitcast is needed?

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 8]
```diff
@@ -109,6 +121,7 @@ def impl_dpnp_empty(
     ty_shape,
     ty_dtype,
     ty_order,
+    # ty_like,
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:124]: remove commented out line

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 9]
```diff
@@ -268,9 +288,11 @@ def codegen(context, builder, sig, llargs):
 @intrinsic
 def impl_dpnp_empty_like(
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:289]: docstring

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 10]
```diff
@@ -410,17 +443,58 @@ def impl_dpnp_ones_like(
...
+    ty_fill_value,
+    ty_dtype,
+    ty_order,
+    ty_like,
+    ty_device,
+    ty_usm_type,
+    ty_sycl_queue,
+    ty_retty_ref,
+):
+    ty_retty = ty_retty_ref.instance_type
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:480]: docstring

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 11]
```diff
@@ -208,6 +209,7 @@ def ol_dpnp_empty(
     shape,
     dtype=None,
     order="C",
+    # like=None, # this gets lost when dpnp.empty() is called outside dpjit
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:212]: remove commented out line?

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 12]
```diff
@@ -729,5 +770,66 @@ def impl(
     else:
         raise errors.TypingError(
             "Cannot parse input types to "
-            + f"function dpnp.empty_like({x}, {dtype}, ...)."
+            + f"function dpnp.empty_like({x1}, {dtype}, ...)."
         )
+
+
+@overload(dpnp.full, prefer_literal=True)
+def ov_dpnp_full(
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:778]: we use the `ol_` prefix for overloads

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 13]
```diff
@@ -729,5 +770,66 @@ def impl(
...
+    shape,
+    fill_value,
+    dtype=None,
+    order="C",
+    like=None,
+    device=None,
+    usm_type=None,
+    sycl_queue=None,
+):
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:787]: docstring

File: numba_dpex/core/runtime/_dpexrt_python.c (hunk 2 above)
Comment (chudur-budur) [reply numba_dpex/core/runtime/_dpexrt_python.c:586]: Done.

File: numba_dpex/core/runtime/_dpexrt_python.c [hunk 14]
```diff
@@ -23,6 +23,28 @@
 #include "_queuestruct.h"
 #include "numba/_arraystruct.h"
 
+/* Debugging facilities - enabled at compile-time */
```
[the remark this answers is not in the document]
Comment (chudur-budur) [reply numba_dpex/core/runtime/_dpexrt_python.c:26]: Removed

File: numba_dpex/core/runtime/_dpexrt_python.c (hunk 3 above)
Comment (chudur-budur) [reply numba_dpex/core/runtime/_dpexrt_python.c:26]: Done.

**Review by diptorupd — APPROVED** [review]
Thank you!

**Pull request merged**